### PR TITLE
Add first draft of new blog post - Code reviews: getting started

### PR DESCRIPTION
**Pull request by thibaudcolas**
Related: https://github.com/CodemateLtd/cookbook/pull/1

This is a first draft of a new article for Codemate's blog. It will be published on the current blog/website, but I thought it could be nice to review it via GitHub.

### CODE REVIEW

File: src/pages/code-reviews-getting-started/index.md [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+>
+> In a software-maintenance organization, 55 percent of one-line maintenance changes were in error before code reviews were introduced. After reviews were introduced, only 2 percent of the changes were in error. When all changes were considered, 95 percent were correct the first time after reviews were introduced. Before reviews were introduced, under 20 percent were correct the first time.
+
+### A great way to learn
+
+Finally, this last resource briefly mentions another reason why I think code reviews should be part of everybody’s toolset: you can learn A LOT from them. They are a great place for the team to share their knowledge and learn:
+
+* How other people would approach and solve a given problem.
+* General stylistic or architectural guidelines to follow for the specific project. Those are defined elsewhere (eg. in a styleguide or architecture principles document) but propagate well via code reviews.
```
Comment (roughike) [inline src/pages/code-reviews-getting-started/index.md]: * s/styleguide/style guide
* s/eg./e.g.

File: src/pages/code-reviews-getting-started/index.md [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+## Resources to get started
+
+To give you a headstart in implementing code reviews for your teams, here are three things you will need:
+
+1.  [Code review guidelines](https://github.com/CodemateLtd/cookbook/blob/master/CONTRIBUTING.md). We usually put those in a `CONTRIBUTING.md` file at the root of the project.
+2.  [A pull request template](https://github.com/CodemateLtd/cookbook/blob/master/.github/PULL_REQUEST_TEMPLATE.md). This is where you state that the team does code reviews, that they will be enforced, and that pull requests should be authored accordingly. Ours is simple but gets the job done.
+3.  [Useful resources about how to do this well](https://github.com/CodemateLtd/cookbook/blob/master/CONTRIBUTING.md). Code reviews are a solved problem, there are great resources out there to take inspiration from.
```
Comment (roughike) [inline src/pages/code-reviews-getting-started/index.md]: > Code reviews are a solved problem, there are great resources out there to take inspiration from.

Feels clumsy, maybe try _...are a solved problem, **and** there are great resources..._

File: src/pages/code-reviews-getting-started/index.md [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+* [Code Reviews: Just Do It](https://blog.codinghorror.com/code-reviews-just-do-it/) by Jeff Atwood, my absolute favorite.
+* The Flutter project has good [code review guidelines](https://flutter.io/design-principles/). I particularly like how they talk about the benefits of code reviews. This is a large-scale open-source project so their particular workflow might not fit perfectly for proprietary work, but still worth having a look.
+* Karumi has really nice pull request guidelines and templates as part of their [project quality assurance guidelines](https://github.com/Karumi/project-quality-assurance). I would really like all the teams I’m part of to have such high-quality guidelines.
+* Finally, Michael Lynch’s two-part article on doing code reviews [like a human](https://mtlynch.io/human-code-reviews-1/) is well worth a read. I think it is probably not necessary to get started, but if your team needs to improve how they communicate as well then this contains good advice.
```
Comment (roughike) [inline src/pages/code-reviews-getting-started/index.md]: * Maybe highlight `doing code reviews like a human` instead of `like a human`? I was going to comment that you missed that awesome article but found this when doing a `Ctrl+F` with the url for that article. :)
* The last sentence is hard to parse, maybe add a comma: _if your team needs to improve how they communicate as well, then this contains good advice_

File: src/pages/code-reviews-getting-started/index.md [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+_My friend [@loicteixeira](https://twitter.com/loicteixeira) does wonderful code reviews. Once he taught me about Python’s [mutable default arguments](https://stackoverflow.com/questions/1132941/least-astonishment-and-the-mutable-default-argument) during a review – they are a great way to learn platform-specific gotchas._
+
+## Code reviews in practice
+
+Sounds great on paper, right? 🌈
+
+### Convince the boss
+
+In practice, the first step is to “convince the boss”. Show them those great [metrics on the effectiveness of code reviews](https://blog.codinghorror.com/code-reviews-just-do-it/). Talk about how reviews improve [quality _in terms of business value_](https://www.codemate.com/quality-in-terms-of-business-value). Their main scepticism might be in the cost of having more people working on an otherwise identical amount of changes – this is a good time to talk about the economics of fixing software defects:
```
Comment (roughike) [inline src/pages/code-reviews-getting-started/index.md]: s/scepticism/skepticism

File: src/pages/code-reviews-getting-started/index.md [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+* Karumi has really nice pull request guidelines and templates as part of their [project quality assurance guidelines](https://github.com/Karumi/project-quality-assurance). I would really like all the teams I’m part of to have such high-quality guidelines.
+* Finally, Michael Lynch’s two-part article on doing code reviews [like a human](https://mtlynch.io/human-code-reviews-1/) is well worth a read. I think it is probably not necessary to get started, but if your team needs to improve how they communicate as well then this contains good advice.
+
+## Do it!
+
+Still sceptical? The best way to know whether code reviews will work for your team is to try them. All of the major hosted Git platforms have pull request features with code reviews built-in, you can try them out and see for yourself how many defects you will catch before they even get merged. At Codemate, a good example of this in practice is [my review of Iiro’s inKino project](https://github.com/roughike/inKino/issues/52). I didn’t catch any bugs, but we certainly all learned a lot 😄 – for me, [Flutter](https://www.codemate.com//considering-flutter), and for him a web developer’s perspective on UI code architecture.
```
Comment (roughike) [inline src/pages/code-reviews-getting-started/index.md]: s/sceptical/skeptical

File: src/pages/code-reviews-getting-started/index.md [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+
+Joel Spolsky puts it neatly:
+
+> [It’s harder to read code than to write it](https://www.joelonsoftware.com/2000/04/06/things-you-should-never-do-part-i/).
+
+John Papa recently gave a great talk about what code readability means in practice:
+
+<iframe width="560" height="315" src="https://www.youtube-nocookie.com/embed/56mETnrByBM" frameborder="0" allow="autoplay; encrypted-media" allowfullscreen></iframe>
+
+If you like numbers, another great article on the topic is [Code Reviews: Just Do It](https://blog.codinghorror.com/code-reviews-just-do-it/) by Jeff Atwood. You should definitely read it. I especially like its excerpts from [Code Complete](https://www.amazon.com/exec/obidos/ASIN/0735619670):
```
Comment (roughike) [inline src/pages/code-reviews-getting-started/index.md]: I feel like "you should definitely read it" doesn't really tell the reader _why_ they should read it or why is it a valuable read.

Suggestions:
* tell the reader _why_ the article is such a good read 
* AND/OR drop the "You should definitely read it" statement altogether

File: src/pages/code-reviews-getting-started/index.md [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+
+_Author: Thibaud Colas, Software Architect at Codemate Ltd._
+
+---
+
+## Why it’s good
+
+Developers tend to think of productivity in terms of “how fast can I write the code for this new feature”. We also tend to say “I write code” when describing our jobs. In practice though, professional software development is a matter of understanding the code that is already there: we simply spend much more time reading than writing. Productivity thus depends more on the overall code’s quality and maintainability, rather than the skills of individual contributors making changes.
+
+Joel Spolsky puts it neatly:
```
Comment (roughike) [inline src/pages/code-reviews-getting-started/index.md]: Who is Joel Spolsky? Why does his opinion matter?

Maybe worth introducing him to the reader.

> Joel Spolsky, founder of Trello and Stack Overflow, puts it neatly:

or maybe:

> Joel Spolsky, a widely respected software engineer with decades of experience, puts it neatly:

File: src/pages/code-reviews-getting-started/index.md [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+
+> [It’s harder to read code than to write it](https://www.joelonsoftware.com/2000/04/06/things-you-should-never-do-part-i/).
```
Comment (roughike) [inline src/pages/code-reviews-getting-started/index.md]: This transition from Joel Spolsky to John Papa feels kinda quick. Before transitioning to the next guy, it would be beneficial to tell some anecdotes here why Joel Spolsky is right, or what was the article about.

I wrote some quick ideas down below. Please feel free to cherry pick pieces, or ditch all of it and use them as inspiration for something completely new!

> In the article above, Joel explains why he feels that rewriting big applications from scratch is never a good idea. He also goes into detail how to combat smelly codebases and what are alternatives to big rewrites.
>
> At the time of writing, Netscape had just shipped a public beta of their brand new major release. Since they decided to do a full rewrite which took them 3 years, their competitors were able to gain leverage and Netscape's market share declined. Joel argues that such big rewrites should never be done, and we should focus on code quality and gradual improvements instead. Although the article is now over 18 years old, it still holds true to this day.
>
> From our anecdotal experiences, if we pay attention to code quality right from the beginning, we'll thank ourselves later. If two persons or more understand and find a piece of code well written now, chances are that it will be more pleasant to maintain later.

File: src/pages/code-reviews-getting-started/index.md [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+
+If you like numbers, another great article on the topic is [Code Reviews: Just Do It](https://blog.codinghorror.com/code-reviews-just-do-it/) by Jeff Atwood. You should definitely read it. I especially like its excerpts from [Code Complete](https://www.amazon.com/exec/obidos/ASIN/0735619670):
+
+> … software testing alone has limited effectiveness – the average defect detection rate is only 25 percent for unit testing, 35 percent for function testing, and 45 percent for integration testing. In contrast, the average effectiveness of design and code inspections are 55 and 60 percent. Case studies of review results have been impressive:
+>
+> In a software-maintenance organization, 55 percent of one-line maintenance changes were in error before code reviews were introduced. After reviews were introduced, only 2 percent of the changes were in error. When all changes were considered, 95 percent were correct the first time after reviews were introduced. Before reviews were introduced, under 20 percent were correct the first time.
+
```
Comment (roughike) [inline src/pages/code-reviews-getting-started/index.md]: I feel like it would be good to have a recap from the quote here.

Some potential key points to drive in:
* automated and functional testing are good, but code reviews are a good complimentary step
* code reviews bumped the correctness of code (whatever that means) from under 20 percent to 95 percent; this is a huge improvement!
* this is maybe because it's easier to spot potential errors if there are multiple pairs of eyes looking at the same code
* statistics back all of this up

File: src/pages/code-reviews-getting-started/index.md [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+Developers tend to think of productivity in terms of “how fast can I write the code for this new feature”. We also tend to say “I write code” when describing our jobs. In practice though, professional software development is a matter of understanding the code that is already there: we simply spend much more time reading than writing. Productivity thus depends more on the overall code’s quality and maintainability, rather than the skills of individual contributors making changes.
+
+Joel Spolsky puts it neatly:
+
+> [It’s harder to read code than to write it](https://www.joelonsoftware.com/2000/04/06/things-you-should-never-do-part-i/).
+
+John Papa recently gave a great talk about what code readability means in practice:
+
+<iframe width="560" height="315" src="https://www.youtube-nocookie.com/embed/56mETnrByBM" frameborder="0" allow="autoplay; encrypted-media" allowfullscreen></iframe>
```
Comment (roughike) [inline src/pages/code-reviews-getting-started/index.md]: I would like to have a short recap about the main points of the video here.

Some key points to consider, don't have to use all of them:

* speaker is John Papa, developer advocate working for Microsoft
* you read code more than you write it, according to Robert C. Martin the read/write ratio is "well over 10:1"
* what makes it worse, like Joel Spolsky said earlier, reading code is harder than writing
* favor fast and effective communication over clever and fast code
* clean, readable code helps to avoid rewrites
* use consistent and meaningful names in code
* self-documenting code is better than commenting everything
* write dirty code that works; then clean it

**Review by roughike — COMMENTED** [review]
Generally looks really cool! 

Writing can be a subjective topic, so I'll just say that here are some of my _opinionated observations_ that came to mind when reading the article.

I also feel like the article introduction block needs some love, but then again I don't have any improvement suggestions so I didn't comment anything there ¯\\_(ツ)_/¯

File: src/pages/code-reviews-getting-started/index.md (hunk 4 above)
Comment (thibaudcolas) [reply src/pages/code-reviews-getting-started/index.md]: Is american english the norm in Finland?

File: src/pages/code-reviews-getting-started/index.md (hunk 6 above)
Comment (thibaudcolas) [reply src/pages/code-reviews-getting-started/index.md]: The reason is "If you like numbers", implying that the article contains statistics. If that's not clear enough I should probably reword, yup.

File: src/pages/code-reviews-getting-started/index.md (hunk 7 above)
Comment (thibaudcolas) [reply src/pages/code-reviews-getting-started/index.md]: I don't think this is worth the extra sentence – either you know who he is already (because he's famous) or you don't and you can click the link which is much more informative than I can ever be.

File: src/pages/code-reviews-getting-started/index.md (hunk 4 above)
Comment (roughike) [reply src/pages/code-reviews-getting-started/index.md]: Ah, my bad! I thought it felt weird and maybe not a word at all; silly me.

File: src/pages/code-reviews-getting-started/index.md (hunk 6 above)
Comment (roughike) [reply src/pages/code-reviews-getting-started/index.md]: In that case, maybe:

> Another great article on the topic is [Code Reviews: Just Do It](https://blog.codinghorror.com/code-reviews-just-do-it/) by Jeff Atwood. If you like numbers, you should definitely read it.

File: src/pages/code-reviews-getting-started/index.md [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+
+## Resources to get started
+
+To give you a headstart in implementing code reviews for your teams, here are three things you will need:
+
+1.  [Code review guidelines](https://github.com/CodemateLtd/cookbook/blob/master/CONTRIBUTING.md). We usually put those in a `CONTRIBUTING.md` file at the root of the project.
+2.  [A pull request template](https://github.com/CodemateLtd/cookbook/blob/master/.github/PULL_REQUEST_TEMPLATE.md). This is where you state that the team does code reviews, that they will be enforced, and that pull requests should be authored accordingly. Ours is simple but gets the job done.
+3.  [Useful resources about how to do this well](https://github.com/CodemateLtd/cookbook/blob/master/CONTRIBUTING.md). Code reviews are a solved problem, there are great resources out there to take inspiration from.
+
+Take all of these, sprinkle some LGTM ([Looks Good To Me](http://knowyourmeme.com/memes/lgtm)) images on top when approving changes, and you’re good to go. Code reviews can feel tense at times, those silly images set a more relaxed tone, and can ease the relationship between reviewers and reviewees. Plus, a nice picture is a great reward for good work!
```
Comment (iiska) [inline src/pages/code-reviews-getting-started/index.md]: This is important even though looking for some nice pics in the middle of the busy things might feel tedious.

Another thing to preserve humane and friendly atmosphere I feel should be mentioned somewhere could go along the lines:
> Be kind and let tooling do the nitpicking. Choose common formatting and enforce it via tooling (Prettier, linters etc). Let humans focus on logic, use cases, readability and learning.
>
> Common formatting makes diffs easier to read and you don't really want to discuss tabs vs spaces on each review.

**Review by iiska — COMMENTED** [review]
Overall I like this article and that one point about mentioning tooling and formatting is something you could decide if it fits in the scope or should be an another story. :smiley:

File: src/pages/code-reviews-getting-started/index.md (hunk 4 above)
Comment (thibaudcolas) [reply src/pages/code-reviews-getting-started/index.md]: 👌 it's probably not a great choice of word for what I'm trying to say, I'll replace it with "concern" (people might have concerns even if they aren't skeptical).

File: src/pages/code-reviews-getting-started/index.md (hunk 5 above)
Comment (thibaudcolas) [reply src/pages/code-reviews-getting-started/index.md]: Fuck I use that word a lot.

File: src/pages/code-reviews-getting-started/index.md (hunk 2 above)
Comment (thibaudcolas) [reply src/pages/code-reviews-getting-started/index.md]: Thank you, will use your wording as-is.

File: src/pages/code-reviews-getting-started/index.md (hunk 3 above)
Comment (thibaudcolas) [reply src/pages/code-reviews-getting-started/index.md]: 👍 that last sentence sucks, I'll try to see if I can get rid of it.

File: src/pages/code-reviews-getting-started/index.md (hunk 6 above)
Comment (thibaudcolas) [reply src/pages/code-reviews-getting-started/index.md]: 👌 edited

File: src/pages/code-reviews-getting-started/index.md (hunk 8 above)
Comment (thibaudcolas) [reply src/pages/code-reviews-getting-started/index.md]: I quite like it as-is, but I can see why if you don't know those people that well it feels weird to name-drop people in that quick of a succession. And it's not fair to expect the audience of this post to be familiar with those names.

I'll reword to rely less on the names, but I won't add new content summarising either of those resources as I think it breaks the flow of this article.

File: src/pages/code-reviews-getting-started/index.md (hunk 9 above)
Comment (thibaudcolas) [reply src/pages/code-reviews-getting-started/index.md]: 👌 good point

File: src/pages/code-reviews-getting-started/index.md (hunk 10 above)
Comment (thibaudcolas) [reply src/pages/code-reviews-getting-started/index.md]: I think the video would be a good watch in its own right, and I don't want to break the flow of the article with too many recaps, so I'll leave this as-is.

File: src/pages/code-reviews-getting-started/index.md (hunk 11 above)
Comment (thibaudcolas) [reply src/pages/code-reviews-getting-started/index.md]: 👌 good point. I'll try to cover this around this "LGTM" part.

**Comment by thibaudcolas** [comment]
Who wants to do another round of reviews, and give this a thumbs up or thumbs down?

Edit: seems like the Netlify integration has stopped working. Here is the rendered post: https://github.com/thibaudcolas/codemate-blog/blob/feature/blog-post-code-reviews-getting-started/src/pages/code-reviews-getting-started/index.md

**Comment by roughike** [comment]
I read it through couple times just now with fresh eyes, and...

g1463594554310385012

**Review by roughike — APPROVED** [review]

**Comment by thibaudcolas** [comment]
Updated to match the latest changes from the published version 🎉 https://www.codemate.com/code-reviews-getting-started/
